feat: CLDSRV-353 processVersioningState() null key support

jonathan-gramain opened the pull request:
Add support for null key handling for the helper
processVersioningState(), and maintain the null version compatibility mode to keep the old behavior - for now, the calling code sets this flag to "true" without using the config to maintain current behavior, it will be changed with CLDSRV-355.

jonathan-gramain committed:
feat: CLDSRV-353 processVersioningState() null key support

Add support for null key handling for the helper
processVersioningState(), and maintain the null version compatibility
mode to keep the old behavior - for now, the calling code sets this
flag to "true" without using the config to maintain current behavior,
it will be changed with CLDSRV-355.

One slight behavior change in compatibility mode is that when an old
null versioned key is deleted due to a PUT overwriting the null
version, we do not send the "replayId" to the DELETE request, but
instead, rely on the "oldReplayId" sent by the PUT, because this is
the normal way of letting metadata know how to get rid of the existing
replayId for the existing null version on versioning-suspended buckets.

jonathan-gramain committed:
CLDSRV-353 [optim] no legacy null version key deletion

In case the master has been updated with a null-key-enabled
Cloudserver, there can be no more versioned key associated (as the new
behavior guarantees that a null master cannot have an associated null
versioned key, see S3C-7526).

Thanks to this, we can avoid having to check the versioned key for
deletion when a null master version is updated on a
versioning-suspended bucket, which is arguably a rather common use
case.

To implement that, we will add a "isNull2" attribute to the master key
in a subsequent commit (part of CLDSRV-355) when Cloudserver is not in
null version compatibility mode.

This commit is implementing the optimization by checking the new
"isNull2" metadata attribute, and skipping the null version check in
case the flag is set.

jonathan-gramain committed:
CLDSRV-353 bump arsenal dep

bert-e commented:
# Hello jonathan-gramain,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-353 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.70.19`

* `8.4.20`

* `8.5.21`

* `8.6.9`

* `8.7.11`


Please check the `Fix Version/s` of CLDSRV-353, or the target
branch of this pull request.

jonathan-gramain committed:
CLDSRV-353 remove conversion of null version key to null key

Fix an issue that occurred when converting a null versioned key into a
null key (that would occur when a non-compat mode Cloudserver updates
a compat-mode object having a noncurrent null version). The issue was
that the null key was updated with the master's contents instead of
the original null version contents.

The fix consists of keeping the backward compatibility by setting a
`nullVersionId` instead, which avoids to have to read the null version
metadata first. It is not important to convert those keys as the
migration from V0 to V1 will necessarily have to convert existing
legacy null versions anyway.

jonathan-gramain commented:
@watsaqat @tmacro I fixed an issue with this PR, could you please review the latest commit https://github.com/scality/cloudserver/pull/5064/commits/c8671ffc7788c437f626c9a3b51395e9566789a1?

jonathan-gramain committed:
CLDSRV-353 + case of delete null versioned key

Add a case in processVersioningState() to delete the null versioned
key where the key is a legacy null version (we know this because
`isNull2` is not set) and we are going to write it as a null key,
because in older Cloudservers there may be an associated null version
as the master null in certain circumstances.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-353 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.70.20`

* `8.4.20`

* `8.5.21`

* `8.6.9`

* `8.7.12`


Please check the `Fix Version/s` of CLDSRV-353, or the target
branch of this pull request.

jonathan-gramain commented:
ping

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.4/feature/CLDSRV-353-modifyPreprocessingVersioningForNullKeys` with contents from `feature/CLDSRV-353-modifyPreprocessingVersioningForNullKeys`
and `development/8.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.4/feature/CLDSRV-353-modifyPreprocessingVersioningForNullKeys origin/development/8.4
 $ git merge origin/feature/CLDSRV-353-modifyPreprocessingVersioningForNullKeys
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.4/feature/CLDSRV-353-modifyPreprocessingVersioningForNullKeys
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/CLDSRV-353-modifyPreprocessingVersioningForNullKeys` into
`development/7.70`
* [w/8.4/feature/CLDSRV-353-modifyPreprocessingVersioningForNullKeys](https://github.com/scality/cloudserver/compare/development/8.4...w/8.4/feature/CLDSRV-353-modifyPreprocessingVersioningForNullKeys) will be merged into `development/8.4` 
* [w/8.5/feature/CLDSRV-353-modifyPreprocessingVersioningForNullKeys](https://github.com/scality/cloudserver/compare/development/8.5...w/8.5/feature/CLDSRV-353-modifyPreprocessingVersioningForNullKeys) will be merged into `development/8.5` 
* [w/8.6/feature/CLDSRV-353-modifyPreprocessingVersioningForNullKeys](https://github.com/scality/cloudserver/compare/development/8.6...w/8.6/feature/CLDSRV-353-modifyPreprocessingVersioningForNullKeys) will be merged into `development/8.6` 
* [w/8.7/feature/CLDSRV-353-modifyPreprocessingVersioningForNullKeys](https://github.com/scality/cloudserver/compare/development/8.7...w/8.7/feature/CLDSRV-353-modifyPreprocessingVersioningForNullKeys) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

jonathan-gramain commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`

* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`

* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`



Please check the status of the associated issue CLDSRV-353.

Goodbye jonathan-gramain.
